compiler/src/dotty/tools/dotc/transform/Splicing.scala

```diff
@@ -198,10 +200,57 @@ class Splicing extends MacroTransform:
       val newTree = transform(tree)
       val (refs, bindings) = refBindingMap.values.toList.unzip
       val bindingsTypes = bindings.map(_.termRef.widenTermRefExpr)
-      val methType = MethodType(bindingsTypes, newTree.tpe)
+      val types = bindingsTypes.collect {
+        case AppliedType(tycon, List(arg: TypeRef)) if tycon.derivesFrom(defn.QuotedTypeClass) => arg
+      }
+      val newTypeParams = types.map { tpe =>
+        newSymbol(
+          spliceOwner,
+          UniqueName.fresh(tpe.symbol.name.toTypeName),
+          Param,
+          TypeBounds.empty
+        )
+      }
+      val methType =
+        if types.nonEmpty then
+          PolyType(types.map(tp => UniqueName.fresh(tp.symbol.name.toTypeName)))(
+            pt => types.map(_ => TypeBounds.empty),
+            pt => {
+               val tpParamMap = new TypeMap {
+                private val mapping = types.map(_.typeSymbol).zip(pt.paramRefs).toMap
+                def apply(tp: Type): Type = tp match
+                  case tp: TypeRef => mapping.getOrElse(tp.typeSymbol, tp)
+                  case tp => mapOver(tp)
+               }
+               MethodType(bindingsTypes.map(tpParamMap), tpParamMap(newTree.tpe))
+            }
+          )
+        else MethodType(bindingsTypes, newTree.tpe)
       val meth = newSymbol(spliceOwner, nme.ANON_FUN, Synthetic | Method, methType)
-      val ddef = DefDef(meth, List(bindings), newTree.tpe, newTree.changeOwner(ctx.owner, meth))
-      val fnType = defn.FunctionType(bindings.size, isContextual = false).appliedTo(bindingsTypes :+ newTree.tpe)
+
+      def substituteTypes(tree: Tree): Tree =
+        if types.nonEmpty then
+          val typeIndex = types.zipWithIndex.toMap
+          TreeTypeMap(
+            typeMap = new TypeMap {
+              def apply(tp: Type): Type = tp match
+                case tp @ TypeRef(x: TermRef, _) if tp.symbol == defn.QuotedType_splice => tp
+                case tp: TypeRef if tp.typeSymbol.hasAnnotation(defn.QuotedRuntime_SplicedTypeAnnot) => tp
+                case tp: TypeRef =>
+                  typeIndex.get(tp) match
+                    case Some(idx) => newTypeParams(idx).typeRef
```

Comment:
Having a Map where the keys are Types might be a little bit fragile since we could have two equivalent Types which are not referentially equal, is this not a problem here?

Reply:
I think the prefix of a TypeRef could be an uncacheable type, for example if it contains a LazyRef